File: lintcode_algo_2021/s4.-queue-bfs-graph/457.Implement_Queue_by_Linked_List/457_implement_queue_by_linked_list__circular_array.py
class MyQueue:
    """
    @param: item: An integer
    @return: nothing
    """
    def __init__(self):
        self.size = 1000000
        self.queue = [0] * self.size
        self.head, self.tail = 0, 0
    def enqueue(self, item):
        # write your code here
        if (self.tail + 1) % self.size == self.head:
            return 
        self.queue[self.tail] = item
        # circular array
        self.tail = (self.tail + 1) % self.size

    """
    @return: An integer
    """
    def dequeue(self):
        # write your code here
        if self.head == self.tail:
            return -1
        item = self.queue[self.head]
        # circular array
        self.head = (self.head + 1) % self.size
        return item
        

File: lintcode_algo_2021/s4.-queue-bfs-graph/457.Implement_Queue_by_Linked_List/457_implement_queue_by_linked_list__circular_array.java
public class MyQueue {
    /*
     * @param item: An integer
     * @return: nothing
     */
    
    public int head, tail;
    public int SIZE = 1000000;
    public int[] queue = new int[SIZE];

    public void enqueue(int item) {
        // write your code here
        if (((tail + 1) % SIZE) == head) {
            return;
        }

        queue[tail] = item;
        tail = (tail + 1) % SIZE;
    } 

    /*
     * @return: An integer
     */
    public int dequeue() {
        // write your code here
        if ((head % SIZE) == tail) {
            return -1000;
        }
        int item = queue[head];
        head = (head + 1) % SIZE;
        return item;
    }
}